New fix for issue 380 Correct recovery of I/O NIC

ShienXie opened the pull request:
Signed-off-by: Shien Xie <shien.xie@kaixiangtech.com>

The fix is to let sheepdog switching back to I/O NIC once I/O NIC is back online.
This is how the fix works:
A. We keep 2 sockfd_cache_fd arrays in each sockfd_cache_entry. The first array holds all fds for I/O NIC and the second holds all fds for Non-I/O NIC. When sheepdog starts, all fds in both arrays are -1 and immediately sheepdog pre-connect all fds. When a sheepdog worker thread (ie: gateway thread) needs a fd, it tries to get one from I/O NIC. If failed, it tries to get from the second array. This guarantees I/O NIC is always being chosen first.
B. A dedicated monitor thread is created to monitor all I/O NICs for each node(entry) in sockfd_cache.root. If the thread found an I/O NIC lost connection, it will close all fds in the first array and set their value to -1. This is to prevent worker thread from using broken fds. If that happens, those thread are typically blocked at network I/O system calls for a long time.
C. Once the monitor thread find the I/O NIC connection is back online, the thread will reconnect all fds in the first array for I/O NIC.
D. The number of fds in both arrays will be increased at the same time, when watermark of either array has been reached.

Comparing to the fix we submitted in April, this fix
1. restrict the changes in sockfd_cache.c. By doing this, functions outside sockfd_cache.c will not be affected. The previous fix we submitted requires other programmers remember the status of the IO ports, which is not very good.
2. use one single thread to detect IO ports for all nodes. The previous change requires a thread for each node.
3. separate sockfd_cache_fd into two arrays and pre-connect all fds in those two arrays. The purposes for this design is to decrease the chance that worker threads blocked in I/O operations for broken network connection and hence, improve the performance. With this design, conceptually, there will be no "short" fds. All fds are "long" fds. None-IO ports can be used if no IO ports are available. If there is no network problem, we don't close any fds.

We tested the fix with a sheepdog cluster of 24 servers.

Link to patch:
https://github.com/sheepdog/sheepdog/pull/406.patch

ShienXie commented:
When I run functional check in house, I don't have any problem. Not sure why Travis CI build failed at test 098. I retested 098 several times but can not reproduce the failure.

mitake commented on lib/sockfd_cache.c:
```diff
@@ -517,3 +637,45 @@ void sockfd_cache_del(const struct node_id *nid, struct sockfd *sfd)
+
+	for (;;) {
+		struct sockfd_cache_entry *entry;
+		rb_for_each_entry(entry, &sockfd_cache.root, rb) {
```
We need to acquire the write lock of `sockfd_cache.lock` here

mitake replied:
Sorry, read lock would be enough here. `prepare_conns()` acquires the write lock.

ShienXie replied:
Not sure why we need to put a read lock here and where should we unlock it. If we unlock after "rb_for_each_entry" block, the sd_write_lock function in prepare_conns() will hang indefinitely.

mitake replied:
This is because other threads can touch `sockfd_cache` e.g. by calling `sockfd_cahe_add()`. The required changes would be:
1. adding a flag for switching acquire or not acquire write lock to `prepare_conns()`
1. let `monitor_sd_node_connectivity()` acquire the write lock of `sockfd_cache.lock`

If it is costly for you, I can do it after merging this version of this PR.

mitake replied:
> Sorry, read lock would be enough here. prepare_conns() acquires the write lock.
 
I mean, the above comment was incorrect (not reader, we need writer), sorry again ;)

ShienXie replied:
@mitake Sure, if you like, you can merge this PR now and make the changes after. I will also look into it once I get time. I think the locking mechanism might be tricky. Sheep may hang if we put a wrong lock. Ie. Cascaded write or Read/Write mix locks.

mitake replied:
OK, then I'll merge your PR this week and add follow up changes. I'll let you know when my changes appear.

mitake commented on lib/sockfd_cache.c:
```diff
@@ -98,12 +101,24 @@ static inline int get_free_slot(struct sockfd_cache_entry *entry)
-		idx = i;
-		break;
+		/*let caller know if this is an IO or NonIO port*/
+		idx = (i + 1) << 20;
```
It seems a little bit hacky. How about let `get_free_slot()` have a new parameter of `enum channel_status*` and notify the type of socket to caller via the value?

mitake commented:
@ShienXie if you don't have a time for resolving the comment, I can do this instead of you. Please let me know.

ShienXie commented:
@mitake Sorry I was very busy recently. I will look into it in next a couple of days.

ShienXie commented:
@mitake I rewrote the change according to your requests. 
1. In order to reduce the time that cache lock is taken by monitor_sd_node_connectivity, I created a linked list so the lock can be released when connect_to_addr() is called.
2. I added an extra parameter to indicate which fds group the free slot belongs to. I did not use 'enum channel_status*' because channel_status means different thing and theoretically, even if IO channel is up, a thread can still get a free slot from NonIO fds group, similar to the short fds of original design.

mitake commented:
@ShienXie merged, thanks a lot! Probably I'll add a cleaning for simplifying locking rule of `to_connect_list`. I'll notify you about the change.

ShienXie commented:
@mitake Great! I am looking forward to see your updates.
